cumulus/client/executor-gossip/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+				);
+				ValidationResult::Discard
+			},
+		}
+	}
+
+	/// Produce a closure for validating messages on a given topic.
+	///
+	/// The gossip engine will periodically prune old or no longer relevant messages using
+	/// `message_expired`.
+	fn message_expired<'a>(&'a self) -> Box<dyn FnMut(Block::Hash, &[u8]) -> bool + 'a> {
+		Box::new(move |_topic, mut data| {
+			let msg_hash = twox_64(data);
+			// TODO: can be expired due to the message itself might be too old?
+			let _msg = match GossipMessage::<Block>::decode(&mut data) {
+				Ok(msg) => msg,
+				Err(_) => return true,
+			};
+			let expired = {
+				let known_rebroadcasted = self.known_rebroadcasted.read();
+				known_rebroadcasted.contains(&msg_hash)
+			};
+			if expired {
+				let mut known_rebroadcasted = self.known_rebroadcasted.write();
+				known_rebroadcasted.remove(&msg_hash);
+			}
+			expired
+		})
+	}
+
+	/// Produce a closure for filtering egress messages.
+	///
+	/// Called before actually sending a message to a peer.
+	fn message_allowed<'a>(
+		&'a self,
+	) -> Box<dyn FnMut(&PeerId, MessageIntent, &Block::Hash, &[u8]) -> bool + 'a> {
+		let do_rebroadcast = {
+			let now = Instant::now();
+			let mut next_rebroadcast = self.next_rebroadcast.lock();
+			if now >= *next_rebroadcast {
+				*next_rebroadcast = now + REBROADCAST_AFTER;
+				true
+			} else {
+				false
+			}
+		};
+
+		Box::new(move |_who, intent, _topic, mut data| {
+			if let MessageIntent::PeriodicRebroadcast = intent {
+				return do_rebroadcast
+			}
+
+			match GossipMessage::<Block>::decode(&mut data) {
+				Ok(_) => true,
+				Err(_) => false,
+			}
+		})
+	}
+}
+
+/// Parameters to run the executor gossip service.
+pub struct ExecutorGossipParams<Block: BlockT, Network, Executor> {
+	/// Substrate network service.
+	pub network: Network,
+	/// Executor instance.
+	pub executor: Executor,
+	/// Stream of transaction bundle produced locally.
+	pub bundle_receiver: TracingUnboundedReceiver<Bundle<Block::Extrinsic>>,
+	/// Stream of execution receipt produced locally.
+	pub execution_receipt_receiver: TracingUnboundedReceiver<ExecutionReceipt<Block::Hash>>,
+}
+
+/// Starts the executor gossip worker.
+pub async fn start_gossip_worker<Block, Network, Executor>(
+	gossip_params: ExecutorGossipParams<Block, Network, Executor>,
+) where
+	Block: BlockT,
+	Network: GossipNetwork<Block> + Send + Sync + Clone + 'static,
+	Executor: GossipMessageHandler<Block> + Send + Sync + 'static,
+{
+	let ExecutorGossipParams { network, executor, bundle_receiver, execution_receipt_receiver } =
+		gossip_params;
+
+	let gossip_validator = Arc::new(GossipValidator::new(executor));
+	let gossip_engine =
+		GossipEngine::new(network, EXECUTOR_PROTOCOL_NAME, gossip_validator.clone(), None);
+
+	let gossip_worker = GossipWorker::new(
+		gossip_validator,
+		Arc::new(Mutex::new(gossip_engine)),
+		bundle_receiver,
+		execution_receipt_receiver,
+	);
+
+	gossip_worker.run().await
```

Comment:
Be careful, AFAIK this will return immediately after the first message. So you'll need something like this instead:
```rust
loop {
    // And this will in turn change to something that races between worker running and receiving messages from the outside eventually
    (&mut gossip_worker).run().await;
}

Reply:
I have tested the code by running the local dev network, I see no issue with the current code and it works great.

Reply:
Hm, maybe I missed something, fine then